lib/key_manager.rb

```diff
@@ -0,0 +1,71 @@
+require 'securerandom'
+require 'thread'
+
+class KeyManager
+  attr_reader :keys, :blocked_keys
+
+  def initialize
+    @keys = {}
+    @blocked_keys = {}
+    @mutex = Mutex.new
+  end
+
+  def generate_key
+    key = SecureRandom.hex(16)
+    # 5 min expiry
+    expiry = Time.now + 300 
+    @mutex.synchronize do
+      @keys[key] = expiry
+    end
+    key
+  end
+
+  def get_available_key
+    @mutex.synchronize do
+      now = Time.now
+      available_keys = @keys.reject { |_, expiry| expiry < now }
+      unblocked_keys = available_keys.reject { |key, _| @blocked_keys.key?(key) }
```

Comment:
Here you are first looping over all the keys available to reject the ones that are expired. 
After that you are looping over all the available to reject blocked keys

In any case you are looping over twice. You will have to optimise this.

Reply:
Have optimised this